src/action/impl.rs

```diff
@@ -0,0 +1,129 @@
+use std::{fmt::Debug, future::Future, pin::Pin};
+
+use serde::{Deserialize, Serialize};
+
+use super::{ActionKind, OperationTarget};
+use crate::{backend::Backend, Gateway};
+
+/// The marker trait for all action runs, this trait should not be used and is only used
+/// to make the return type of [`Gateway::run`] easily known.
+///
+/// [`Gateway::run`]: crate::Gateway::run
+pub trait ActionRunner<Success, Failure>: private::Sealed + Send {
+	#[doc(hidden)]
+	unsafe fn __run<B: Backend>(
```

Comment:
This should also be unhidden and allow users to do both `Gateway::run` and `Action::run`